NugetLightRequest.cs

```diff
@@ -1755,40 +1907,69 @@ private IEnumerable<PackageItem> SearchForPackages(PackageSource source, string
                         tempName = string.Empty;
                     }
 
                     searchTerm = tempName;
                 }
 
+                if (!String.IsNullOrWhiteSpace(searchTerm))
+                {
+                    searchTerms.Add(new NuGetSearchTerm(NuGetSearchTerm.NuGetSearchTermType.SearchTerm, searchTerm));
+                    searchTermsDebugMessageBuilder.AppendFormat(CultureInfo.InvariantCulture, "searchTerm:{0}", searchTerm);
+                }
+
                 // Add the Tags for the search.
                 if (FilterOnTag != null)
                 {
                     // E.g. searchTerm = "tag:dscresource_xFirefox tag:command_Start-Process"
-                    searchTerm = FilterOnTag.Value.Where(tag => !string.IsNullOrWhiteSpace(tag)).Aggregate(searchTerm, (current, tag) => current + " tag:" + tag);
+                    foreach (string tag in FilterOnTag.Value.Where(tag => !string.IsNullOrWhiteSpace(tag)))
+                    {
+                        searchTerms.Add(new NuGetSearchTerm(NuGetSearchTerm.NuGetSearchTermType.Tag, tag));
+                        searchTermsDebugMessageBuilder.AppendFormat(CultureInfo.InvariantCulture, " tag:{0}", tag);
+                    }
+                }
+
+                // Add Contains term for search
+                if (!string.IsNullOrWhiteSpace(Contains.Value))
+                {
+                    searchTerms.Add(new NuGetSearchTerm(NuGetSearchTerm.NuGetSearchTermType.Contains, Contains.Value));
+                    searchTermsDebugMessageBuilder.AppendFormat(CultureInfo.InvariantCulture, " contains:\"{0}\"", Contains.Value);
                 }
 
-                Verbose(Resources.Messages.SearchingRepository, source.Repository.Source, searchTerm);
+                Verbose(Resources.Messages.SearchingRepository, source.Repository.Source, searchTermsDebugMessageBuilder.ToString());
 
                 // Handling case where a user does not provide Name parameter
-                var pkgs = source.Repository.Search(searchTerm, this);
+                NuGetSearchContext searchContext = new NuGetSearchContext()
+                {
+                    SearchTerms = searchTerms,
+                    AllowPrerelease = AllowPrereleaseVersions.Value,
+                    AllVersions = AllVersions.Value
+                };
+                NuGetSearchResult searchResult = source.Repository.Search(searchContext, this);
+                var pkgs = searchResult.Result;
 
-                if (!String.IsNullOrWhiteSpace(name))
+                if (!String.IsNullOrWhiteSpace(name) && searchResult.NamePostFilterRequired)
                 {
                     //Filter on the results. This is needed because we replace [...] regex in the searchterm at the begining of this method.
-                    pkgs = FilterOnName(pkgs, name, isNameContainsWildCard);
+                    FilterEntryByName nameFilter = new FilterEntryByName();
```

Comment:
is there possible we can make it static to avoid new everytime?

Reply:
44f4229